course_catalog/models.py

```diff
@@ -187,3 +185,21 @@ class ProgramItem(ListItem):
     """
 
     program = models.ForeignKey(Program, related_name="items", on_delete=models.CASCADE)
+
+
+class FavoriteItem(TimestampedModel):
+    """
+    FavoriteItem model tracks LearningResources that are marked by user as favorite.
+    Favorites don't need to track an user-specified order, although they can be
+    displayed ordered by timestamp. Users should be able to favorite an
+    LearningResource, including other other Lists like Programs and UserLists.
+    """
+
+    user = models.ForeignKey(User, on_delete=models.PROTECT)
+    content_type = models.ForeignKey(
+        ContentType,
+        limit_choices_to={"model__in": ("Course", "Bootcamp", "UserList", "Program")},
+        on_delete=models.CASCADE,
+    )
+    object_id = models.PositiveIntegerField()
+    item = GenericForeignKey("content_type", "object_id")
```

Comment:
There should probably be a unique constraint across the tuple of `('user', 'content_type', 'object_id')` to ensure an item can't be favorited multiple times.

Reply:
Sure thing.